htsohm/pseudomaterial_generator/random.py

```diff
@@ -0,0 +1,138 @@
+from random import random
+
+from Crypto.Hash import SHA256
+from zlib import crc32
+
+from htsohm.db import Material
+from htsohm.structure import Structure, LatticeConstants, AtomSite, AtomType
+
+# random number generator
+def bytes_to_float(b):
+    return float(crc32(b) & 0xffffffff) / 2**32
+
+def str_to_float(s, encoding="utf-8"):
+    return bytes_to_float(s.encode(encoding))
+
+def random_number(n):
+    # use SHA256 to get random number from some input value
+    h = SHA256.new('{}'.format(n).encode())
+    return str_to_float(h.hexdigest())
+
+# uniform selection from random number
+def uniform_selection(a, b, r):
+    return b - r * (b - a)
+
+def random_number_density(number_density_limits, structure, r): #lattice_constants):
+    """Produces random number for atom-sites in a unit cell, constrained by
+    some number density limits.
+
+    Args:
+        number_density_limits (list): min and max number densities, for example:
+            [min_(float), max_(float)]
+        lattice_constants (dict): crystal lattice constants, for example:
+            {"a" : (float),
+             "b" : (float),
+             "c" : (float)}
+    
+    Returns:
+        atoms (int): some random number of atom-sites under the imposed limits.
+
+        If the minimum number density results in a unit cell with less than 2
+        atom-sites with the given lattice constants, a minimum number density
+        of TWO ATOM SITES PER UNIT CELL is imposed.
+    
+    """
+    min_ND = number_density_limits[0]
+    max_ND = number_density_limits[1]
+    v = structure.volume()
+    min_atoms = int(min_ND * v)
+    max_atoms = int(max_ND * v)
+    if min_atoms < 2:
+        min_atoms = int(2)
+    return int(uniform_selection(min_atoms, max_atoms, r))
+
+def generate_material(run_id, seed, config):
+    """Create records for pseudomaterial simulation and structure data."
+
+    Args:
+        run_id (str): identification string for run.
+        number_of_atomtypes (int): number of different chemical species used to
+            populate the unit cell.
+
+    Returns:
+        material (sqlalchemy.orm.query.Query): database row for storing 
+            simulation data specific to the material. See
+            `htsohm/db/material.py` for more information.
+
+    """
+    number_of_atom_types    = config["number_of_atom_types"]
+    lattice_limits          = config["lattice_constant_limits"]
+    number_density_limits   = config["number_density_limits"]
+    epsilon_limits          = config["epsilon_limits"]
+    sigma_limits            = config["sigma_limits"]
+    max_charge              = config["charge_limit"]
+
+    ########################################################################
+    # create database row
+    material = Material(run_id)
+    material.seed = seed
+
+    # initialize incrementable value for random number generation
+    seed
```

Comment:
should this be doing something else here?

Reply:
this is leftover code from when i was initializing the seed value within this function, i'd moved that step up one level, but i just added comments so it's more clear.

see:

commit 4d648c538c339be56486584051e7a42f08a77cc5
Author: akaija <a.r.kaija@gmail.com>
Date:   Wed Jun 20 16:12:01 2018 -0400

    added comments explaining where seed is initialized